sim/battle.ts

```diff
@@ -74,6 +75,22 @@ type Part = string | number | boolean | Pokemon | Side | Effect | Move | null |
 // An individual Side's request state is encapsulated in its `activeRequest` field.
 export type RequestState = 'teampreview' | 'move' | 'switch' | '';
 
+type EventHandlerParam<T, U extends 0 | 1 | 2 | 3, Default = never> = Exclude<T, undefined> extends never ? Default :
+	Exclude<T, string | number | boolean | undefined> extends infer R ?
+		R extends (...args: any) => any ?
+			Parameters<R>[U] extends infer S ?
+				Exclude<S, undefined> extends never ?
+					Default :
+					undefined extends S ?
+						S | null :
+						S :
+				never :
+			never :
+		never;
```

Comment:
I'll be honest I'm looking at this and similar types and I'm thoroughly confused on what its supposed to do, which concerns me as I'm sure other developers will be in the same boat. This needs to be documented. This can include a comment above the typecheck and you might want to make your generic type names clearer (not a single letter) since those can be any identifier. I know its standard practice to use a capital letter but I think we can make an exception here for readability's sake.

Ties in with L449-515